File: java/com/joshmanisdabomb/lcc/block/network/BlockNetwork.java
package com.joshmanisdabomb.lcc.block.network;

import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public abstract class BlockNetwork<T> {

    private final int distance;

    public BlockNetwork(int distance) {
        this.distance = distance;
    }

    public NetworkResult discover(World world, T start) {
        List<T> traversables = new ArrayList<>(Collections.singletonList(start));
        List<BlockPos> nodes = new ArrayList<>();
        BlockPos startPos = this.toPosition(start);
        for(int i = 0; i < traversables.size(); i++) {
            T traversable = traversables.get(i);
            for (T otherTraversable : this.traverse(world, traversable, nodes)) {
                BlockPos bp = this.toPosition(otherTraversable);
                if ((startPos.equals(bp) || startPos.withinDistance(bp, this.distance)) && !traversables.contains(otherTraversable)) {
                    traversables.add(otherTraversable);
                }
            }
        }
        return this.createNetwork(traversables, nodes.stream().distinct().collect(Collectors.toList()));
    }

    protected abstract List<T> traverse(World world, T current, List<BlockPos> nodes);

    protected abstract BlockPos toPosition(T traversable);

    protected NetworkResult createNetwork(List<T> traversables, List<BlockPos> nodes) {
        return new NetworkResult(traversables, nodes);
    }

    public class NetworkResult {

        private final List<T> traversables;
        private final List<BlockPos> nodes;

        private NetworkResult(List<T> traversables, List<BlockPos> nodes) {
            this.traversables = traversables;
            this.nodes = nodes;
        }

        @Override
        public String toString() {
            return this.getClass().getSimpleName() + ": " + this.traversables.size() + " traversables, " + this.nodes.size() + " nodes";
        }

        public BlockPos toPosition(T traversable) {
            return BlockNetwork.this.toPosition(traversable);
        }

        public List<T> getTraversables() {
            return new ArrayList<>(this.traversables);
        }

        public List<BlockPos> getTraversablePositions() {
            return this.traversables.stream().map(this::toPosition).distinct().collect(Collectors.toList());
        }

        public List<BlockPos> getNodePositions() {
            return new ArrayList<>(this.nodes);
        }

    }

}

File: lcc-network/src/main/kotlin/com/joshmanisdabomb/lcc/network/BlockNetwork.kt
package com.joshmanisdabomb.lcc.network

import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.BlockView

abstract class BlockNetwork<T>(val distance: Int) {

    protected val directions = Direction.values()

    fun discover(world: BlockView, start: T): NetworkResult {
        val traversables = mutableListOf(start)
        val nodes = mutableMapOf<String, MutableSet<T>>()
        val startPos = toPosition(start)
        var i = 0
        while (i < traversables.size) {
            val t = traversables[i]
            val others = traverse(world, t, nodes)
            for (other in others) {
                val bp = toPosition(other)
                if (!traversables.contains(other) && (startPos == bp || startPos.isWithinDistance(bp, distance.toDouble()))) {
                    traversables.add(other)
                }
            }
            i++
        }
        return result(traversables.distinct().toSet(), nodes)
    }

    protected abstract fun traverse(world: BlockView, current: T, nodes: MutableMap<String, MutableSet<T>>): Set<T>

    protected abstract fun toPosition(traversable: T): BlockPos

    protected open fun result(traversables: Set<T>, nodes: Map<String, Set<T>>) = NetworkResult(traversables, nodes)

    protected fun from(nodes: MutableMap<String, MutableSet<T>>, type: String) = nodes.computeIfAbsent(type) { mutableSetOf() }

    inner class NetworkResult internal constructor(private val _traversables: Set<T>, private val _nodes: Map<String, Set<T>>) {

        override fun toString() = "${this::class.simpleName}: ${traversables.size} traversables, ${nodes.size} nodes"

        fun toPosition(traversable: T) = this@BlockNetwork.toPosition(traversable)

        val traversables get() = _traversables.toList()
        val nodes get() = _nodes.toMap()

        val traversablesAssoc get() = traversables.groupBy(this::toPosition).mapValues { (_, v) -> v.distinct() }
        val nodesAssoc get() = nodes.mapValues { (_, v) -> v.groupBy(this::toPosition).mapValues { (_, v) -> v.distinct() } }

    }

}